web/src/main/java/com/google/accompanist/web/WebView.kt

```diff
@@ -0,0 +1,189 @@
+                        url: String?,
+                        isReload: Boolean
+                    ) {
+                        super.doUpdateVisitedHistory(view, url, isReload)
+                        // WebView will often update the current url itself.
+                        // This happens in situations like redirects and navigating through
+                        // history. We capture this change and update our state holder url.
+                        // On older APIs (28 and lower), this method is called when loading
+                        // html data. We don't want to update the state in this case as that will
+                        // overwrite the html being loaded.
+                        if (url != null &&
+                            !url.startsWith("data:text/html") &&
+                            state.content.getCurrentUrl() != url
+                        ) {
+                            state.content = WebContent.Url(url)
+                        }
+                    }
+
+                    override fun onReceivedError(
+                        view: WebView?,
+                        request: WebResourceRequest?,
+                        error: WebResourceError?
+                    ) {
+                        super.onReceivedError(view, request, error)
+                        onError(request, error)
+                    }
+
+                    override fun shouldOverrideUrlLoading(
+                        view: WebView?,
+                        request: WebResourceRequest?
+                    ): Boolean {
+                        // Override all url loads to make the single source of truth
+                        // of the URL the state holder Url
+                        request?.let {
+                            val content = WebContent.Url(it.url.toString())
+                            state.content = content
+                        }
+                        return true
+                    }
+                }
+            }.also { webView = it }
+        },
+        modifier = modifier
+    ) { view ->
+        when (val l = state.content) {
+            is WebContent.Url -> {
+                val url = l.url
+
+                if (url.isNotEmpty() && url != view.url) {
+                    view.loadUrl(url)
+                }
+            }
+            is WebContent.Data -> {
+                view.loadDataWithBaseURL(l.baseUrl, l.data, null, "utf-8", null)
+            }
+        }
+
+        canGoBack = view.canGoBack()
+    }
+}
+
+sealed class WebContent {
+    data class Url(val url: String) : WebContent()
+    data class Data(val data: String, val baseUrl: String? = null) : WebContent()
+
+    fun getCurrentUrl(): String? {
+        return (this as? Url)?.url
+    }
+}
+
+/**
+ * A state holder to hold the state for the WebView. In most cases this will be remembered
+ * using the rememberWebViewState(uri) function.
+ */
+class WebViewState(webContent: WebContent) {
```

Comment:
To check with eng, but I think this class should be annotated with `@Stable`. I'm not sure if the compiler is able to infer that `content` and `isLoading` are fields using `mutableStateOf` APIs.

Reply:
Pretty sure you are right so I have added it.